### PR TITLE
[`pylint`] Implement `empty-comment` (`PLR2044`)

### PR DESCRIPTION
**Pull request by mdbernard**
## Summary

Part of #970.

This adds Pylint's [R0244 empty_comment](https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/empty-comment.html) lint as well as an always-safe fix.

## Test Plan

The included snapshot verifies the following:

- A line containing only a non-empty comment is not changed
- A line containing leading whitespace before a non-empty comment is not changed
- A line containing only an empty comment has its content deleted
- A line containing only leading whitespace before an empty comment has its content deleted
- A line containing only leading and trailing whitespace on an empty comment has its content deleted
- A line containing trailing whitespace after a non-empty comment is not changed
- A line containing only a single newline character (i.e. a blank line) is not changed
- A line containing code followed by a non-empty comment is not changed
- A line containing code followed by an empty comment has its content deleted after the last non-whitespace character
- Lines containing code and no comments are not changed
- Empty comment lines within block comments are ignored
- Empty comments within triple-quoted sections are ignored

## Comparison to Pylint

Running Ruff and Pylint 3.0.3 with Python 3.12.0 against the `empty_comment.py` file added in this PR, we see the following:

* Identical behavior:
    * empty_comment.py:3:0: R2044: Line with empty comment (empty-comment)
    * empty_comment.py:4:0: R2044: Line with empty comment (empty-comment)
    * empty_comment.py:5:0: R2044: Line with empty comment (empty-comment)
    * empty_comment.py:18:0: R2044: Line with empty comment (empty-comment)

* Differing behavior:
    * Pylint doesn't ignore empty comments in block comments commonly used for visual spacing; I decided these were fine in this implementation since many projects use these and likely do not want them removed.
        * empty_comment.py:28:0: R2044: Line with empty comment (empty-comment)
    * Pylint detects "empty comments" within the triple-quoted section at the bottom of the file, which is arguably a bug in the Pylint implementation since these are not truly comments. These are ignored by this implementation.
        * empty_comment.py:37:0: R2044: Line with empty comment (empty-comment)
        * empty_comment.py:38:0: R2044: Line with empty comment (empty-comment)
        * empty_comment.py:39:0: R2044: Line with empty comment (empty-comment)

### CODE REVIEW

**Comment by mdbernard** [comment]
Looks like the ecosystem checks found some edge cases not currently covered:

- "empty comment" lines within block comments, which add vertical whitespace for clarity, especially common as header comments for files
- lines containing a trailing `#` character in multi-line comments, which should not be considered empty comments

**Comment by charliermarsh** [comment]
Nice, thanks for kicking this off! Agreed on the edge cases described above.

File: crates/ruff_linter/resources/test/fixtures/pylint/empty_comment.py
Comment (mdbernard) [inline crates/ruff_linter/resources/test/fixtures/pylint/empty_comment.py:1]: Add tests for "empty comment" lines in multi-line strings/comments and in block comments

File: crates/ruff_linter/src/rules/pylint/rules/empty_comment.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+/// PLR2044
+pub(crate) fn empty_comment(line: &Line) -> Option<Diagnostic> {
+    if let Some((column, last_non_whitespace_char)) = line
+        .as_str()
+        .char_indices()
+        .rev()
+        .find(|&(_, c)| !is_python_whitespace(c))
+    {
+        if last_non_whitespace_char == '#' {
+            #[allow(clippy::cast_possible_truncation)]
```
Comment (mdbernard) [inline crates/ruff_linter/src/rules/pylint/rules/empty_comment.rs:53]: - [ ] Justify safety

File: crates/ruff_linter/src/rules/pylint/rules/empty_comment.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            let start = TextSize::new((line.start().to_usize() + column) as u32);
+            let last = match line
+                .as_str()
+                .char_indices()
+                .rev()
+                .find(|&(_, c)| !is_python_whitespace(c) && c != '#')
+            {
+                Some((i, _second_to_last_non_whitespace_char)) =>
+                {
+                    #[allow(clippy::cast_possible_truncation)]
```
Comment (mdbernard) [inline crates/ruff_linter/src/rules/pylint/rules/empty_comment.rs:63]: - [ ] Justify safety

**Comment by mdbernard** [comment]
- [ ] Account for comment style with trailing `#` at end of comment, e.g. https://github.com/rotki/rotki/blob/63302ca6d67ab928fd6cac7b6137c41df416416c/rotkehlchen/chain/ethereum/modules/l2/loopring.py#L527:

```python
# -- Methods following the EthereumModule interface -- #
```

**Comment by mdbernard** [comment]
- [ ] Account for comment style with multiple `#` symbols on a single line (tying in with the above comment, probably best to just lint when a comment consists of only a single `#` symbol followed by only whitespace through the end of the line.

File: crates/ruff_linter/src/rules/pylint/rules/empty_comment.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        return None; // the comment is not empty
+    }
+
+    let last = match line
+        .as_str()
+        .char_indices()
+        .rev()
+        .find(|&(_, c)| !is_python_whitespace(c) && c != '#')
+    {
+        #[allow(clippy::cast_possible_truncation)] // TODO: justify safety
```
Comment (mdbernard) [inline crates/ruff_linter/src/rules/pylint/rules/empty_comment.rs:70]: - [ ] Justify safety

File: crates/ruff_linter/src/rules/pylint/rules/empty_comment.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        Some((i, _second_to_last_non_whitespace_char)) => {
+            TextSize::new((line.start().to_usize() + i + 1) as u32)
+        }
+        None => line.start(),
+    };
+
+    Some(
+        Diagnostic::new(
+            EmptyComment,
+            #[allow(clippy::cast_possible_truncation)] // TODO: justify safety
```
Comment (mdbernard) [inline crates/ruff_linter/src/rules/pylint/rules/empty_comment.rs:80]: - [ ] Justify safety

File: crates/ruff_python_trivia/src/comment_ranges.rs [hunk 5]
```diff
@@ -45,6 +48,98 @@ impl CommentRanges {
...
+    /// # This is not a block comment because
+    ///     # not all consecutive lines have the
+    /// # first `#` character in the same column
+    ///
+    /// """
+    /// # This is not a block comment because it is
+    /// # contained within a multi-line string/comment
+    /// """
+    /// ```
+    pub fn block_comments(&self, locator: &Locator) -> Vec<Vec<TextSize>> {
```
Comment (mdbernard) [inline crates/ruff_python_trivia/src/comment_ranges.rs:82]: - [ ] Try to simplify

**Review by mdbernard — COMMENTED** [review]
- [ ] Rebase

**Comment by mdbernard** [comment]
- [ ] Add comparison to Pylint R2044

**Review by charliermarsh — COMMENTED** [review]
This looks great, I really like what I'm seeing in the ecosystem checks too -- thank you!

I'm going to explore making this a token-based rule rather than a physical line-based rule, just because we want to avoid using physical lines when possible, but it'll be a mostly mechanical change...

**Comment by charliermarsh** [comment]
> Pylint doesn't ignore empty comments in block comments commonly used for visual spacing; I decided these were fine in this implementation since many projects use these and likely do not want them removed.

I agree, we shouldn't flag these.

> Pylint detects "empty comments" within the triple-quoted section at the bottom of the file, which is arguably a bug in the Pylint implementation since these are not truly comments. These are ignored by this implementation.

I agree, flagging empty comments within a multi-line string seems like a bug to me.

**Comment by mdbernard** [comment]
Noting that I might've missed an edge case where two empty comments exist on consecutive lines forming a "block" comment.

**Comment by mdbernard** [comment]
> we want to avoid using physical lines when possible

Out of curiosity, why?

**Comment by charliermarsh** [comment]
> Out of curiosity, why?

Primarily because physical line-based approaches tend to be slower, since we need to run the rule over _every_ line. This rule is actually a good example: we already _know_ the locations of all comments in the file (via `CommentRanges`), so iterating over every line, then checking if the line intersects with the comment ranges, is wasteful as compared to merely iterating over the comment ranges directly. In general, we often have information we've already computed and/or can exploit to limit the search space vs. iterating over the physical lines.

**Review by charliermarsh — APPROVED** [review]
This is great work! I ended up moving to the token checker, and added support for the case in which we have multiple consecutive empty comments.

**Pull request merged**